iss/generator.py

```diff
@@ -194,3 +203,315 @@ def to_fastq(generator, output):
             for read_tuple in generator:
                 SeqIO.write(read_tuple[0], f, "fastq-sanger")
                 SeqIO.write(read_tuple[1], r, "fastq-sanger")
+
+
+def worker_iterator(work, error_model, cpu_number, worker_prefix, seed, sequence_type, gc_bias):
+    """A utility function to run the reads simulation of each record in a loop for a specific cpu"""
+    logger = logging.getLogger(__name__)
+    try:
+        forward_handle = open(f"{worker_prefix}_R1.fastq", "w")
+        reverse_handle = open(f"{worker_prefix}_R2.fastq", "w")
+    except PermissionError as e:
+        logger.error("Failed to temporary output file(s): %s" % e)
+        sys.exit(1)
+
+    if seed is not None:
+        random.seed(seed + cpu_number)
+        np.random.seed(seed + cpu_number)
+
+    with forward_handle, reverse_handle:
+        for record, n_pairs, mode in work:
+            simulate_reads(
+                record=record,
+                n_pairs=n_pairs,
+                mode=mode,
+                error_model=error_model,
+                cpu_number=cpu_number,
+                forward_handle=forward_handle,
+                reverse_handle=reverse_handle,
+                sequence_type=sequence_type,
+                gc_bias=gc_bias,
+            )
+
+
+def generate_work_divider(
+    fasta_file, readcount_dic, abundance_dic, n_reads, coverage, coverage_file, error_model, output, chunk_size
+):
+    """Yields a list of tuples containing the records and the number of reads to generate for each record
+
+    Yields:
+        list[tuple[SeqIO.Record, int, str]]: a list of tuples containing the records,
+            the number of reads to generate for each record and the mode
+    """
+    logger = logging.getLogger(__name__)
+
+    current_chunk = 0
+    total_reads_generated = 0
+    total_reads_generated_unrounded = 0
+
+    chunk_work = []
+
+    for record in fasta_file:
+        # generate reads for records
+        if readcount_dic is not None:
```

Comment:
any reason you are using `is not None` here and in a few others places instead of simply `if readcount_dic:` ?

Reply:
Not really, the readcount_dic is either None or a non-empty dict `: dict[] | None`, so it would evaluate the same. (Adding type hinting at some point would be awesome! :D)  
I'll change it to `is not None` everywhere, because "Explicit is better than Implicit"